File: splearn/src/test/java/jp/kopher1601/splearn/application/member/required/MemberRepositoryTest.java
package jp.kopher1601.splearn.application.member.required;

import jakarta.persistence.EntityManager;
import jp.kopher1601.splearn.domain.member.Member;
import jp.kopher1601.splearn.domain.member.MemberStatus;
import lombok.RequiredArgsConstructor;
import org.junit.jupiter.api.Test;
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
import org.springframework.dao.DataIntegrityViolationException;
import org.springframework.test.context.TestConstructor;

import static jp.kopher1601.splearn.domain.member.MemberFixture.createMemberRegisterRequest;
import static jp.kopher1601.splearn.domain.member.MemberFixture.createPasswordEncoder;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.assertj.core.api.AssertionsForClassTypes.assertThat;

@DataJpaTest
@TestConstructor(autowireMode = TestConstructor.AutowireMode.ALL)
@RequiredArgsConstructor
class MemberRepositoryTest {

    private final MemberRepository memberRepository;
    private final EntityManager entityManager;

    @Test
    void createMember() {
        Member member = Member.register(createMemberRegisterRequest(), createPasswordEncoder());

        assertThat(member.getId()).isNull();

        memberRepository.save(member);

        assertThat(member.getId()).isNotNull();

        entityManager.flush();
        entityManager.clear();

        Member found = memberRepository.findById(member.getId()).orElseThrow();
        assertThat(found.getStatus()).isEqualTo(MemberStatus.PENDING);
        assertThat(found.getDetail().getRegisteredAt()).isNotNull();
    }

    @Test
    void duplicateEmailFail() {
        Member member1 = Member.register(createMemberRegisterRequest(), createPasswordEncoder());
        memberRepository.save(member1);

        Member member2 = Member.register(createMemberRegisterRequest(), createPasswordEncoder());
        assertThatThrownBy(() -> memberRepository.save(member2))
            .isInstanceOf(DataIntegrityViolationException.class);
    }
}

File: spring-exposed/src/test/kotlin/jp/kopher/springexposed/repository/MemberRepositoryTest.kt
package jp.kopher.springexposed.repository

import jp.kopher.springexposed.domain.Member
import jp.kopher.springexposed.domain.Members
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.test.context.TestConstructor
import org.springframework.transaction.annotation.Transactional

@SpringBootTest
@Transactional
@TestConstructor(autowireMode = TestConstructor.AutowireMode.ALL)
class MemberRepositoryTest(
    private val memberRepository: MemberRepository,
) {

    @Test
    fun `会員を保存して ID を返す`() {
        val member = Member(
            name = "Exposed",
            age = 2,
        )

        // when
        val resultRows = memberRepository.save(member)

        assertThat(resultRows[Members.id]).isNotNull()
        assertThat(resultRows[Members.name]).isEqualTo("Exposed")
        assertThat(resultRows[Members.age]).isEqualTo(2)
    }

    @Test
    fun `ID を指定して会員を検索できる`() {
        val member = Member(
            name = "Kotlin",
            age = 2,
        )
        val savedMember = memberRepository.save(member)

        // when
        val foundMember = memberRepository.findById(savedMember[Members.id])

        assertThat(foundMember).isNotNull()
        assertThat(foundMember?.name).isEqualTo("Kotlin")
        assertThat(foundMember?.age).isEqualTo(2)
    }
}